lib/land/trackers/api_tracker.rb

```diff
@@ -206,24 +199,74 @@ def user_agent
                      Land.config.blank_user_agent_string
 
         @user_agent = UserAgent[user_agent]
+        @user_agent.user_agent_type = UserAgentType['user']
 
         if Land.config.identify_crawlers && defined?(CrawlerDetect)
           crawler_detect = CrawlerDetect.new(user_agent)
-          user_agent_type = crawler_detect.is_crawler? ? 'crawl' : 'api'
-          @user_agent.update(user_agent_type: UserAgentType[user_agent_type])
+          @user_agent.user_agent_type = UserAgentType['crawl'] if crawler_detect.is_crawler?
         end
 
         browser = ::Browser.new(user_agent)
+        land_browser = Land::Browser[browser.name]
 
-        @user_agent.update(
-          browser: Browser[browser.name],
-          device: Device[browser.device.name],
-          platform: Platform[browser.platform.name],
-          browser_version: browser.version
-        )
+        @user_agent.browser = land_browser
+        @user_agent.device = Device[browser.device.name]
+        @user_agent.platform = Platform[browser.platform.name]
+        @user_agent.browser_version = browser.version
+        @user_agent.device_resolution = device_resolution
+        @user_agent.browser_color_preference = browser_color_preference
+
+        @user_agent.save! if @user_agent.changed?
+
+        @user_agent
+      rescue ActiveRecord::RecordNotUnique
+        retry
+      rescue ActiveRecord::RecordInvalid => e
+        retry if e.message == 'Validation failed: User agent has already been taken'
 
+        add_error_tag_to_land_span(e)
         @user_agent
       end
+
+      def device_resolution
+        return @device_resolution if defined?(@device_resolution)
+
+        device_width = params.dig('device_resolution', 'width')
+        device_height = params.dig('device_resolution', 'height')
+        device_orientation = params.dig('device_resolution', 'orientation')
+
+        @device_resolution = if device_width && device_height && device_orientation
+                               DeviceResolution.find_or_create_by(
+                                 device_resolution: "#{device_width}x#{device_height}",
+                                 width: device_width,
+                                 height: device_height,
+                                 orientation: device_orientation
+                               )
+                             end
```

Comment:
Logical check should be an early return guard, `UserAgent#save!` should be responsible for making the save (`lookup_by` is rough here, as it is doing a `find_or_create_by` effectively, though we will recrify this when removing `lookup_by` 

```suggestion
         next unless device_width && device_height && device_orientation
         
        @device_resolution = DeviceResolution.find_or_initialize_by(
                               device_resolution: "#{device_width}x#{device_height}",
                               width: device_width,
                               height: device_height,
                               orientation: device_orientation
                            )
  ```

Reply:
I cant seem to use next unless here, because is is not in a block or loop! Since it is in a method instead I will have to use return unless!